processing/src/main/java/org/apache/druid/segment/RowWalker.java

```diff
@@ -60,13 +61,16 @@ public T currentRow()
 
   public void advance()
   {
-    if (rowIterator.hasNext()) {
+    if (rowIterator == null) {
+      throw new IllegalStateException("cannot call advance when isDone == true");
+    } else if (rowIterator.hasNext()) {
       current = rowIterator.next();
 
       if (current == null) {
         throw new NullPointerException("null row encountered in walker");
       }
     } else {
+      rowIterator = null;
```

Comment:
nit: perhaps better do `assert rowIterator == null` in `isDone()`.

Reply:
Makes sense to me, but that's gonna get changed in #11886 anyway.